packages/babel-types/src/validators/isBlockScoped.ts

```diff
@@ -1,10 +1,20 @@
-import { isClassDeclaration, isFunctionDeclaration } from "./generated";
+import {
+  isClassDeclaration,
+  isFunctionDeclaration,
+  isImportDeclaration,
+} from "./generated";
 import isLet from "./isLet";
 import type * as t from "..";
 
 /**
  * Check if the input `node` is block scoped.
  */
 export default function isBlockScoped(node: t.Node): boolean {
-  return isFunctionDeclaration(node) || isClassDeclaration(node) || isLet(node);
+  return (
+    isFunctionDeclaration(node) ||
+    isClassDeclaration(node) ||
+    isLet(node) ||
+    // import declaration can be block scoped when it's in TS/flow module declaration
+    isImportDeclaration(node)
```

Comment:
Actually it is not block scoped as `ImportDeclaration`s may only appear in the top level or inside a module/namespace (for TS). Adding it here may be a breaking change for someone using the validator.
Furthermore excluding `ImportDeclaration`s from https://github.com/babel/babel/blob/fce35af69101c6b316557e28abf60bdbf77d6a36/packages/babel-traverse/src/scope/index.ts#L201-L212
results in excluding those nodes from the traversal and not running `registerDeclaration` on them.

Reply:
I don't think excluding `ImportDeclarations` is well enough, since it will also cause breaking changes.

The main problem here is that **the scope of declared module should be isolated with other modules' scopes and the program's scope**. But unfortunately it's not now.

So maybe we could change the implementation of `getProgramParent` to make it returns not only on `Program`, but also `ModuleDeclaration`. This will make scopes of modules isolated with each other  and the program.

Reply:
Yep, excluding `ImportDeclaration` is definitely not the right way here.
You will be happy to know that there is already a method called `getBlockParent` that gets the upper block. However I would suggest you to give a look at how the `BlockScoped` visitor is handling block scoped declarations, because we may have to replicate a similar behavior for `ImportDeclaration`.

Reply:
> make it returns not only on Program, but also ModuleDeclaration.

I don't think we can mark ModuleDeclaration as program because variables within module declaration can reference bindings defined in program:

```js
type T = number;
declare module 'test' {
  declare var a: T;
}
```

Since ambient declarations are always top level. We can check if `node.parent` is `ModuleDeclaration` and then register declarations there, otherwise we fallback to 

https://github.com/babel/babel/blob/fce35af69101c6b316557e28abf60bdbf77d6a36/packages/babel-traverse/src/scope/index.ts#L209-L210

Reply:
I took a different approach.

Take all `ImportDeclaration` into a separated visitor, and use `getBlockParent` to find parent scope.

I think it is more conducive to maintenance。